policyeval/broker.go

```diff
@@ -0,0 +1,371 @@
+	return &Broker{
+		logger:           l.Named("broker"),
+		nackTimeout:      timeout,
+		deliveryLimit:    deliveryLimit,
+		pendingEvals:     make(map[string]PendingEvaluations),
+		enqueuedEvals:    make(map[string]int),
+		enqueuedPolicies: make(map[string]string),
+		unack:            make(map[string]*unackEval),
+		waiting:          make(map[string]chan struct{}),
+	}
+}
+
+// Enqueue adds an eval to the broker.
+func (b *Broker) Enqueue(eval *sdk.ScalingEvaluation) {
+	b.l.Lock()
+	defer b.l.Unlock()
+	b.enqueueLocked(eval, "")
+}
+
+func (b *Broker) enqueueLocked(eval *sdk.ScalingEvaluation, token string) {
+	logger := b.logger.With(
+		"eval_id", eval.ID, "policy_id", eval.Policy.ID,
+		"queue", eval.Policy.Type, "token", token)
+
+	logger.Debug("enqueue eval")
+
+	// Check if eval is already enqueued.
+	if _, ok := b.enqueuedEvals[eval.ID]; ok {
+		if token == "" {
+			logger.Debug("eval already enqueued")
+			return
+		}
+	} else {
+		b.enqueuedEvals[eval.ID] = 0
+	}
+
+	queue := eval.Policy.Type
+
+	// Get pending heap for the policy type.
+	pending, ok := b.pendingEvals[queue]
+	if !ok {
+		// Initialize pending heap and waiting channel.
+		pending = PendingEvaluations{}
+		if _, ok := b.waiting[queue]; !ok {
+			b.waiting[queue] = make(chan struct{}, 1)
+		}
+	}
+
+	// Check if an eval for the same policy is already enqueued.
+	pendingEvalID, ok := b.enqueuedPolicies[eval.Policy.ID]
+	if !ok {
+		b.enqueuedPolicies[eval.Policy.ID] = eval.ID
+	} else if pendingEvalID != eval.ID {
+		logger.Debug("policy already enqueued")
+
+		// Policy is waiting to be evaluated, but this could be a newer
+		// evaluation request and the policy could have changed. So update
+		// the pending heap with the new eval.
+		i, pendingEval := pending.GetEvaluation(pendingEvalID)
+		if pendingEval != nil {
+			if eval.CreateTime.After(pendingEval.CreateTime) {
+				logger.Debug("new eval is newer, policy updated")
+				b.enqueuedPolicies[eval.Policy.ID] = eval.ID
+				delete(b.enqueuedEvals, eval.ID)
+				pending[i] = eval
+				heap.Fix(&pending, i)
+			}
+			return
+		}
+	}
+
+	heap.Push(&pending, eval)
+	b.pendingEvals[queue] = pending
+
+	// Unblock any blocked dequeues.
+	select {
+	case b.waiting[queue] <- struct{}{}:
+	default:
+	}
+
+	logger.Debug("eval enqueued")
+}
+
+// Dequeue is used to retrieve an eval from the broker.
+func (b *Broker) Dequeue(ctx context.Context, queue string) (*sdk.ScalingEvaluation, string, error) {
+	logger := b.logger.With("queue", queue)
+
+	logger.Debug("dequeue eval")
+
+	eval := b.findWork(queue)
+	for eval == nil {
+		proceed := b.waitForWork(ctx, queue)
+		if !proceed {
+			return nil, "", nil
+		}
+
+		eval = b.findWork(queue)
+	}
+
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Generate an UUID for the token.
+	token := uuid.Generate()
+
+	// Setup Nack timer.
+	// Eval needs to be Ack'd before this timer finishes.
+	nackTimer := time.AfterFunc(b.nackTimeout, func() {
+		b.Nack(eval.ID, token)
```

Comment:
Do we need to handle the error returned by `Nack` here for at least logging purposes?

Reply:
agreed. the nack/ack errors seem to fall into two categories:
* don't bother retrying (eval doesn't exist, token is wrong, already nack'd), but something may be wrong.
* don't need to do anything (in which case, maybe shouldn't have returned an error)